### ISSUE TITLE
Failed network plugins should set NodeNetworkUnavailable

### ISSUE THREAD

**Issue by squeed** [kubernetes/kubernetes#56794]
**Is this a BUG REPORT or FEATURE REQUEST?**:
> /kind feature

**What happened**:
When the network plugin detects a failure (e.g. the CNI configuration is not on disk), we don't set NodeNetworkUnavailable. Instead, the kubelet sets NodeReady to false with a comma-separated list of all errors as the Message.

This is hard to parse out, and often masks the true problem. It also is confusing to end-users, because the state NodeNetworkUnavailable only means that a PodCIDR hasn't been assigned.

**What you expected to happen**:
A network failure causes NodeNetworkUnavailable to be set

**How to reproduce it (as minimally and precisely as possible)**:
Bring up a kubelet with `--network-plugin=cni` and no CNI configuration files on disk.

**Comment by s-urbaniak** [comment]
To add more context to it, `NodeNetworkUnavailable` currently is set mainly in the node controller range allocation only: https://github.com/kubernetes/kubernetes/blob/v1.8.4/pkg/controller/node/ipam/sync/sync.go#L343.

Hence the current semantic meaning of `NodeNetworkUnavailable` only relates to PodCIDR allocations as explained by @squeed above.

**Comment by cmluciano** [comment]
@kubernetes/sig-network-bugs

**Comment by squeed** [comment]
I've been told by @dcbw that "there's a reason for this," but I'm eagerly awaiting what it is :-).

**Comment by caseydavenport** [comment]
> NodeNetworkUnavailable

Am I mistaken in believing that this condition is meant to represent the node's networking status, not status of networking for pods on that node? Obviously these are often tied, but for example on a one node cluster you don't necessarily need the node's networking to be ready if all the pods are going to talk to each other on the same node.

I suspect we need something more fine-grained than this, and a more extensible way of specifying what features a given node supplies to pods (a unique IP address being one, reachability to other nodes being another).

**Comment by caseydavenport** [comment]
I believe there is an ongoing effort to represent these states as taints, so that pods can tolerate conditions that they're OK with.

As a rough strawman, we could create some "well-known" taints as part of this effort which represent various networking conditions on the node - "network plugin ready", "node->node network ready", etc. Then, for example, Pods running in host-network could tolerate the "network plugin ready" and not need CNI to be installed.

**Comment by squeed** [comment]
It seems I was a bit confused, but for good reason - the parameter space is convoluted and non-orthogonal.

NodeNetworkUnavailable is a NodeConditionType. There are 6 possible NodeConditionTypes:
- NodeReady
- NodeOutOfDisk
- NodeMemoryPressure
- NodeDiskPressure
- NodeNetworkUnavailable
- NodeConfigOK (??)

By default, NodeNetworkUnavailable is true in the following cases.
- The IPAM controller has yet to assign an PodCIDR
- A cloud provider is enabled, that cloud provider is `gce`, and no routes have been configured

The four "negative" conditions are also automatically converted to taints.

From a scheduling perspective, `NodeReady` and `NodeNetworkUnavailable` are treated identically. Some of the controller code uses the NodeConditions, and some use the implicit taints - those are equivalent. I only see an effort to remove use of NodeConditions within the scheduler, not in the kubelet. I might be wrong.

There is special code in the kubelet (not in the scheduler!) that fails pods without `hostNetwork: true` if there are any network pugin errors, but not if `NodeNetworkUnavailable` is true. This is in a very bizarre place; it's not an admission controller, it's a "runtime" execution error.

There are some people who manually set NodeNetworkUnavailable to `false` in the apiserver directly, because they are using GCE but not GCE cloud routing. This also needs to be fixed.

In conclusion, I think it would be safe to connect the mapping between `pkg/kubelet/runtimeState.networkError` and NodeNetworkUnavailable. It certainly would be a user-friendly action.

**Comment by jryberg** [comment]
@squeed, Hi, How do you manually set NodeNetworkUnavailable to false in the apiserver?

Regards Johan

**Comment by squeed** [comment]
@jryberg you can just PATCH the Node object.

**Comment by jryberg** [comment]
@squeed do you have any example how to do this patch?

**Comment by cmluciano** [comment]
@caseydavenport @squeed Is it sufficient to push this to the CNI plugin itself or should this condition be set if we discovered that there should be a CNI plugin but there isn't?

**Comment by squeed** [comment]
@cmluciano This definitely doesn't belong in either the CNI plugin or the CNI kubernetes plugin. It should be a part of the kubelet ready logic, where everything else is.

**Comment by caseydavenport** [comment]
Yeah, I don't think this belongs in the CNI plugin itself. It raises a question though of how do we get a full picture of network readiness, since many CNI plugins also rely on one or more daemons running on the node.

Perhaps that's up the each plugin to figure out, or perhaps the kubelet needs some sort of readiness discovery mechanism to check that the requisite bits of code are available / running on that node.

For example, as a strawman, I could imagine something similar to how readiness is exposed for pods today - a list of exec / http endpoints which the kubelet can call to determine if plugins are ready.

**Comment by squeed** [comment]
While making the CNI interface richer is definitely on my list, we currently have an undocumented standard that writing a CNI configuration file to disk signifies network readiness. This PR is just about hooking that up with the already existing network status, instead of node status.

**Comment by caseydavenport** [comment]
Yes, sorry. Didn't mean to try and hijack :)

**Comment by dcbw** [comment]
So NodeNetworkUnavailable originally came from cloud providers and routes as you've found.  The problem is that you may wish to use a cloud provider, but you might *not* use cloud routes at all because you have an overlay.  Cloud routes are mostly relevant for the direct IP routing between nodes eg with flannel + hostgw or maybe other IP-based plugins.

What happened was that if a cloud provider was enabled the node would by default have NodeNetworkUnavailable set which was expected to be cleared by the Route Controller when cloud routes were added.  But if you run an overlay, you don't need and will never add cloud routes, and thus the Route Controller never cleared the condition because it wasn't even run.

NodeNetworkUnavailable conflated (a) having a cloud provider and (b) needing cloud routes for the data plane.  (b) depends on (a), but (a) does not depend on (b).

Perhaps we keep NodeNetworkUnavailable as-is since stuff is supposed to move to taints anyway.

**Comment by dcbw** [comment]
> do you have any example how to do this patch?

@jryberg https://github.com/openshift/origin/blob/master/pkg/network/master/subnets.go#L138

**Comment by dcbw** [comment]
> In conclusion, I think it would be safe to connect the mapping between pkg/kubelet/runtimeState.networkError and NodeNetworkUnavailable. It certainly would be a user-friendly action.

@squeed not quite safe.  You miss the case where there is a network plugin (like flannel + hostgw or kubenet) *and* you require cloud routes to operate the plugin.  In that case, the readiness of both pieces (flannel+hostgw and cloud routes) is a separate thing and if both use NodeNetworkUnavailable, you don't know when to clear the condition.

But we've kinda already go this handled with NodeReady=false-because-network-plugin-not-ready and NodeNetworkUnavailable=true-when-you-have-no-cloud-routes.  But it could certainly be clarified, or perhaps there really is a better scheme and I'm certainly all ears.

Ideally the "do I need cloud routes to operate the data plane" decision would be part of the network plugin, but that significantly complicates network plugins like kubenet or flannel+hostgw and makes them kubernetes specific.  So we decided to make that a cluster admin install problem, such that:

```
    if networkPlugin==kubenet|flannel|etc && cloudProvider=true {
        --configure-cloud-routes=true
    } else {
        --configure-cloud-routes=false
    }
```

eg you as the cluster admin know whether your cluster is on GCE/AWS and whether the network plugin you've selected requires cloud routes to successfully set up the pod data plane.
